### PR TITLE
8913 vd special resolution uxa feedback updates

### PR DESCRIPTION
**Pull request by argush3**
*Issue #:* /bcgov/entity#8913

*Description of changes:*

* UXA feedback updates
* Added tests for `CompleteResolution` component
* misc cleanup

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/assets/styles/base.scss [hunk 1]
```diff
@@ -142,6 +142,10 @@ p {
   font-weight: normal;
 }
 
+.tool-tip {
+  cursor: default
```
Comment (severinbeauvais) [inline src/assets/styles/base.scss:146]: Did UX ask for all tooltips to be an arrow instead of a pointing hand?

May I recommend the "help" cursor for tooltips?

Comment (argush3) [reply src/assets/styles/base.scss:146]: this is what was requested by UX:

>don't change the mouse cursor to the text tool on hover for tool tips (keep as regular pointer)

File: src/components/Affidavit/Affidavit.vue [hunk 2]
```diff
@@ -452,4 +452,11 @@ ul {
     text-decoration: none;
   }
 }
+
+#upload-affidavit-card {
+    // text area word count counter
+    ::v-deep label,v-label,theme--light {
```
Comment (severinbeauvais) [inline src/components/Affidavit/Affidavit.vue:458]: This looks like incorrect syntax. I think maybe you mean:
`::v-deep label.v-label.theme--light`

Comment (severinbeauvais) [reply src/components/Affidavit/Affidavit.vue:458]: "v-label" and "theme-light" are not HTML elements; they are classes, so they need a period in front of them.

This CSS targets all "label"s. Is it working correctly?

File: src/components/CreateResolution/CompleteResolution.vue [hunk 3]
```diff
@@ -96,15 +97,15 @@
...
-      <p class="mt-2">
+      <p class="section-description mt-2">
         {{getCreateResolutionResource.resolutionDateSection.description}}
       </p>
       <div :class="{ 'invalid-section': getShowErrors && !this.isResolutionDateValid }">
         <v-card flat id="resolution-date-card" class="pt-7 pb-5 pl-6 pr-6">
           <v-row>
             <v-col id="resolution-date-card-left-col" cols="2" class="pt-6" >
               <v-card-title class="resolution-date-vcard-title pl-1 pr-0">
-                <div>Resolution Date</div>
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolution.vue:107]: I agree; these divs are needed.

No need to change, but sometimes I wrap title text or button text in spans (eg, when they are on multiple lines) so that we can select the text directly in the unit tests and then use `toBe` instead of `toContain` (due to extra line breaks in the latter).

Comment (severinbeauvais) [reply src/components/CreateResolution/CompleteResolution.vue:107]: ^^ oops, typo. Added "not".

File: src/components/Affidavit/Affidavit.vue (hunk 2 above)
Comment (argush3) [reply src/components/Affidavit/Affidavit.vue:458]: yeah, it seems to be working properly but only because we probably aren't using any other labels outside of the use case I am looking at.  I'll take another look at this.

File: src/components/CreateResolution/CompleteResolution.vue [hunk 4]
```diff
@@ -367,6 +369,11 @@ export default class CompleteResolution extends Mixins(CommonMixin, DateMixin, E
   // Validation Rules
   readonly Rules = Rules
 
+  /** Is True if Jest is running the code. */
+  get isJestRunning (): boolean {
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolution.vue:373]: This is already declared in Common Mixin.

File: src/assets/styles/base.scss (hunk 1 above)
Comment (severinbeauvais) [reply src/assets/styles/base.scss:146]: OK.

Can you ask them if the "help" cursor would be better?

File: src/components/CreateResolution/CompleteResolution.vue [hunk 5]
```diff
@@ -690,6 +714,12 @@ ul {
   color: $gray7;
 }
 
+// Used for title/placeholder text of input fields.  This was required as some input fields require
+// were not using the expected color of $gray7
+::v-deep label,.v-label,theme--light {
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolution.vue:719]: Check classes again.

File: src/components/CreateResolution/CompleteResolution.vue [hunk 6]
```diff
@@ -653,6 +662,21 @@ export default class CompleteResolution extends Mixins(CommonMixin, DateMixin, E
     this.updateResolutionStepValidationDetail()
   }
 
+  private async reRenderResolutionText (): Promise<void> {
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolution.vue:665]: Not async. Does the method within need to be awaited?

Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue:665]: good catch, it used to be awaited.

File: src/components/CreateResolution/CompleteResolutionSummary.vue [hunk 7]
```diff
@@ -155,6 +168,22 @@ export default class CompleteResolutionSummary extends Mixins(DateMixin) {
...
+    this.resolutionTextHeightUpdateRequired = false
+    this.$refs.resolutionTextRef.calculateInputHeight()
+  }
+
+  // Previously, the resolution text area would not render to the appropriate height relative to the amount of content
+  // when navigating from another step.  In hooking into the visibility change event on the resolution text area via the
+  // v-observe-visibility property, we are able to force a re-calculation of the text area height when a user navigates
+  // to the complete resolution summary step from another step for the first time. This results in the text area being
+  // rendered to the appropriate height.
+  private async onResolutionVisibilityChanged (isVisible, entry) {
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolutionSummary.vue:182]: Also not async :)

I'm curious to see this in action. Looks neat.

File: src/components/CreateResolution/CompleteResolution.vue [hunk 8]
```diff
@@ -653,6 +662,21 @@ export default class CompleteResolution extends Mixins(CommonMixin, DateMixin, E
...
+    this.$refs.resolutionTextRef.calculateInputHeight()
+  }
+
+  // Previously, the resolution text area would not render to the appropriate height relative to the amount of content
+  // when navigating from another step.  In hooking into the visibility change event on the resolution text area via the
+  // v-observe-visibility property, we are able to force a re-calculation of the text area height when a user navigates
+  // to the complete resolution step from another step for the first time. This results in the text area being rendered
+  // to the appropriate height.
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolution.vue:673]: Thanks for the comment. This is exactly the kind of thing that needs a comment 👍 

So the component resizes when the user lands on it, or when the page renders?

Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue:673]: it resizes when the component is actually on the complete resolution step and the component is visible in the users view port

File: src/components/CreateResolution/CompleteResolution.vue (hunk 6 above)
Comment (severinbeauvais) [reply src/components/CreateResolution/CompleteResolution.vue:665]: Also `onResolutionVisibilityChanged()` below.

File: src/components/CreateResolution/CompleteResolutionSummary.vue [hunk 9]
```diff
@@ -129,6 +137,11 @@ export default class CompleteResolutionSummary extends Mixins(DateMixin) {
     return '(Not Entered)'
   }
 
+  @Watch('getCreateResolutionStep.resolutionText')
+  private onResolutionTextChanged (): void {
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolutionSummary.vue:141]: Can you just watch `this.resolutionText`, instead of the field on the getter?

Comment (severinbeauvais) [reply src/components/CreateResolution/CompleteResolutionSummary.vue:141]: Or could you add a method to the `@change` event of resolutionTextRef?

Comment (argush3) [reply src/components/CreateResolution/CompleteResolutionSummary.vue:141]: true, let me test that.

File: src/components/CreateResolution/CompleteResolution.vue [hunk 10]
```diff
@@ -653,6 +662,21 @@ export default class CompleteResolution extends Mixins(CommonMixin, DateMixin, E
     this.updateResolutionStepValidationDetail()
   }
 
+  private async reRenderResolutionText (): Promise<void> {
+    this.$refs.resolutionTextRef.calculateInputHeight()
+  }
```
Comment (severinbeauvais) [inline src/components/CreateResolution/CompleteResolution.vue:667]: This is such a small method, and only called from 1 place. Could its body just be moved to the caller?

(And same on CompleteResolutionSummary page.)

File: tests/jest-wrapper-factory.ts [hunk 11]
```diff
@@ -11,13 +13,19 @@ import { getVuexStore } from '@/store'
 // Utils
 import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
 import { IncorporationResources } from '@/resources'
+// This polyfill(intersection-observer) was installed as a dev dependency as without it components being tested
+// that use the IntersectionObserver API results in warnings that browser does not support this API.
+// e.g. IntersectionObserver API is not available in your browser.
+//      Please install this polyfill: https://github.com/w3c/IntersectionObserver/tree/master/polyfill
+require('intersection-observer')
```
Comment (severinbeauvais) [inline tests/jest-wrapper-factory.ts:20]: Thanks for the comment. (It answers the question "why is this here?")

Why use "require"? Does "import" work as well?

File: src/components/CreateResolution/CompleteResolution.vue (hunk 10 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue:667]: yes

File: tests/jest-wrapper-factory.ts (hunk 11 above)
Comment (argush3) [reply tests/jest-wrapper-factory.ts:20]: i think i was just looking at the documentation and just used what was provided.  Will try importing like we do normally.

File: src/assets/styles/base.scss (hunk 1 above)
Comment (argush3) [reply src/assets/styles/base.scss:146]: sure, will ask

Comment (argush3) [reply src/assets/styles/base.scss:146]: reply from Tracey:

> Tracey Lebedovich @tracey.lebedovich
I think just keep it simple and don’t change the cursor.
aka - keep it as a regular pointer cursor
(not text cursor, not help icon)

Comment (severinbeauvais) [reply src/assets/styles/base.scss:146]: ok

File: src/components/CreateResolution/CompleteResolution.vue (hunk 4 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue]: updated

File: src/components/CreateResolution/CompleteResolutionSummary.vue (hunk 9 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolutionSummary.vue:141]: updated to watch `resolutionText`.  Tried `@change` but things weren't updating properly.

File: tests/jest-wrapper-factory.ts (hunk 11 above)
Comment (argush3) [reply tests/jest-wrapper-factory.ts]: updated to use `import` from `require`

File: src/components/CreateResolution/CompleteResolution.vue (hunk 6 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue]: removed this function and moved logic into onResolutionVisibilityChanged

File: src/components/CreateResolution/CompleteResolutionSummary.vue (hunk 7 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolutionSummary.vue]: removed `async`.  yeah, it's kinda cool...this directive.  I could see it coming in handy for other stuff.

File: src/components/CreateResolution/CompleteResolution.vue (hunk 10 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue]: Removed both instances of `reRenderResolutionText` function and moved logic into `onResolutionVisibilityChanged` function.

File: src/components/CreateResolution/CompleteResolutionSummary.vue (hunk 7 above)
Comment (severinbeauvais) [reply src/components/CreateResolution/CompleteResolutionSummary.vue]: Could you do a small tech talk on it?

Comment (argush3) [reply src/components/CreateResolution/CompleteResolutionSummary.vue]: sure, i'll talk about it a bit at the BCROS developer demo thing.

File: src/components/Affidavit/Affidavit.vue (hunk 2 above)
Comment (argush3) [reply src/components/Affidavit/Affidavit.vue]: updated to use correct syntax and tested to be applying styles correctly.

File: src/components/CreateResolution/CompleteResolution.vue (hunk 5 above)
Comment (argush3) [reply src/components/CreateResolution/CompleteResolution.vue]: updated to use correct syntax and tested to be applying styles correctly.

**Review by severinbeauvais — APPROVED** [review]
Looks good,
